### ISSUE TITLE
Implement Back Propagation Algorithm

### ISSUE THREAD

**Issue by kirushyk** [kirushyk/le#19]
We want Neural Networks to report gradients for Gradient Descent optimizers to use.

**Comment by kirushyk** [comment]
Done
